services/port_manager/src/main/java/com/futurewei/alcor/portmanager/service/implement/PortServiceImpl.java

```diff
@@ -247,6 +290,68 @@ private void updatePortToHost(PortState portState) {
 
     }
 
+    private void updatePortStateAsync(PortState portState, PortState oldPortState, AsyncExecutor executor,
+                                      Stack<PortStateRollback> rollbacks) throws Exception {
+        //Update mac_address
+        String macAddress = portState.getMacAddress();
+        String oldMacAddress = oldPortState.getMacAddress();
+        if (macAddress != null && !oldMacAddress.equals(macAddress)) {
+            MacManagerProxy macManagerProxy = new MacManagerProxy(rollbacks);
+            executor.runAsync(macManagerProxy::releaseMacAddress, oldPortState);
```

Comment:
We should allocate mac address before releasing the old one. Otherwise, it is possible that after releasing the old mac, Port Manager instance crashes which causes loss of old mac but no new mac is allocated. In the worse case, the old mac address is allocated by some other tenant.

Reply:
Since both allocate and release are asynchronous, the order of allocate and release cannot be guaranteed.

Reply:
Good point. But we still needs to address this issue. Two options:
* Based on the different exceptions, Port Manager handles it differently. For example, if allocate fails and release succeeds, we need to retry allocation heavily; if allocate succeeds and release fails, we need to verify that the mac is still own by existing port then return new mac.
* MacManager provides one API to "replace" mac for a port in one "transaction". This appears to be cleaner and easier for clients such as PortManager.

What do you think of?

Reply:
Mac Manager to provide an update interface would be nice, at least reducing the number of interactions with it. But this exceptions will still exist(for example, allocate  mac address succeeded, but allocate ip address failed), if we handle this exception in the main flow of CRUD processing, it may be a bit complicated, and there is no 100% guarantee that the exception will be handled correctly (for example, what should we do if we try again but fails?). The simple handling is that once an exception occurs, the whole operation is considered to have failed, and the rollback mechanism is used to undo all operations before the failure.

Reply:
> Mac Manager to provide an update interface would be nice, at least reducing the number of interactions with it. But this exceptions will still exist(for example, allocate mac address succeeded, but allocate ip address failed), 

Does this comment apply to the replacement API? The replacement API offers atomic operation, either success together or failure together, no mix of the two. 

> if we handle this exception in the main flow of CRUD processing, it may be a bit complicated, and there is no 100% guarantee that the exception will be handled correctly (for example, what should we do if we try again but fails?). The simple handling is that once an exception occurs, the whole operation is considered to have failed, and the rollback mechanism is used to undo all operations before the failure.

Correct. The rollback mechanism is intact to protect resources across microservices as described, but not within the same microservices therefore we need the upstream service to provide "replacement" API so that client don't need to worry about those complicated rollback mechanism.

Reply:
If no objection, let us proceed to provide two "replacement" APIs, one in Mac Manager, and one in IP Manager then. Thoughts?

Reply:
Agreed. No problem.

Reply:
Tracked by Issue https://github.com/futurewei-cloud/alcor/issues/229.